polar-core/src/vm.rs

```diff
@@ -1286,8 +1300,8 @@ impl PolarVirtualMachine {
     /// Sort applicable rules by specificity.
     /// Create a choice over the applicable rules.
     fn query_for_predicate(&mut self, predicate: Predicate) -> PolarResult<()> {
-        match self.get_generic_rule(&predicate.name) {
-            None => self.push_goal(Goal::Backtrack)?,
+        let goals = match self.kb.read().unwrap().rules.get(&predicate.name) {
```

Comment:
This is slightly worse to read than the previous version. But I think the only alternative would be to create a new wrapper type that holds the RwLock and derefs to predicate.